python/tvm/relay/op/dyn/_transform.py

```diff
@@ -81,3 +84,23 @@ def _reshape_shape_func_input_data(data, newshape, ndim):
 @_reg.register_shape_func("dyn.reshape", True)
 def dynamic_reshape_shape_func(attrs, inputs, out_ndims):
     return [_reshape_shape_func_input_data(*inputs, out_ndims[0])]
+
+
+@script
+def _tile_shape_func(data, reps, ndim):
+    out = output_tensor((ndim,), "int64")
+
+    for i in const_range(ndim):
+        out[i] = data.shape[i] * int64(reps[i])
+    return out
+
+
+@_reg.register_shape_func("dyn.tile", True)
+def tile_shape_func(attrs, inputs, _):
+    """
+    Shape function for tile op.
+    """
+    ndim = len(inputs[0].shape)
+    rdim = inputs[1].shape[0].value
+    assert ndim == rdim, "tile data and reps ranks don't match"
```

Comment:
Last question, I notice that the shape_func here is somehow different from the previous impl 

https://github.com/apache/incubator-tvm/blob/151f3f5a00b5b2e2a369b9169dc8cda02e87e82a/python/tvm/relay/op/_transform.py#L609

is it expected?

Reply:
Yes, to prevent dynamic rank behavior, I matched the TF/ONNX APIs and only allow dynamic tile where the rank of the input and the size of the reps match. This is slightly different than the standard TOPI op which implicitly pads either the input shape or the reps with ones.

Reply:
I'm thinking about that more this morning, I'm not sure it's 100% required to prevent dynamic ranks, I'm wondering if I can expand it back to cover the other definition.

Reply:
I've been looking at it again this morning, I'm trying to convince myself I might be able to re-implement the MXNet-like behavior without dynamic ranks.

Reply:
I think I got it to work. I wasn't able to use the infrastructure from the static case directly, but it now implements the same API.

Reply:
Cool. Thanks.